stdlib/queue.mli

```diff
@@ -20,6 +20,99 @@
    {b Warning} This module is not thread-safe: each {!Queue.t} value
    must be protected from concurrent access (e.g. with a [Mutex.t]).
    Failure to do so can lead to a crash.
+
+   A basic example:
+    {[
+    # let q = Queue.create ();;
+    val q : '_weak1 Queue.t = <abstr>
+
+
+    # Queue.push 1 q; Queue.push 2 q; Queue.push 3 q;;
+    - : unit = ()
+
+    # Queue.length q;;
+    - : int = 3
+
+    # Queue.pop q;;
+    - : int = 1
+
+    # Queue.pop q;;
+    - : int = 2
+
+    # Queue.pop q;;
+    - : int = 3
+
+    # Queue.pop q;;
+    Exception: Stdlib.Queue.Empty.
+    ]}
+
+   For a more elaborate example, a classic algorithmic use of queues
+   is to implement a BFS (breadth-first search) through a graph.
+
+   {[
+     type graph = {
+       edges: (int, int list) Hashtbl.t
+     };;
+
+    (* Search in graph [g] using BFS, starting from node [start].
+       It returns the first node that satisfies [p], or [None] if
+       no node reachable from [start] satifies [p].
+    *)
+    let search_for ~(g:graph) ~(start:int) (p:int -> bool) : int option =
+      let to_explore = Queue.create() in
+      let explored = Hashtbl.create 16 in
+
+      Queue.push start to_explore;
+
+      let rec loop () =
+        if Queue.is_empty to_explore then None
+        else
+          (* node to explore *)
+          let node = Queue.pop to_explore in
+          explore_node node
+
+      and explore_node node =
+        if p node then Some node (* found *)
+        else (
+
+          if not (Hashtbl.mem explored node) then (
+            Hashtbl.add explored node ();
```

Comment:
I would check `explored` before calling `p`, as the latter may be arbitrarily expensive so we don't want to call it several times on the same node.